File: Java/src/Hangman.java
import java.io.*;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;
import java.util.Random;
import java.util.stream.Stream;

public class Hangman {
    private final List<String> dict = new ArrayList<>();
    private String word = "";
    private String wordDisplay = "";
    private int tries = 6;
    private final Random generator = new Random();


    public Hangman() throws FileNotFoundException {
        loadDict();
    }

    public void start() throws IOException {
        selectWord();
        initWordDisplay();
        printStatus();
        
        Scanner input = new Scanner(System.in);
        displayWord();

        while (this.tries > 0) {
            System.out.print("Make your guess: ");
            String guess = input.nextLine();
            System.out.println();

            if (guess.isBlank() || guess.isEmpty()) {
                System.out.println("You entered nothing, please write something");
            }
            else if (guess.length() == 1) {
                if (fillBlanks(guess.charAt(0))) {
                    System.out.println(guess.charAt(0) + " was in the word, nice!");
                } else {
                    System.out.println(guess.charAt(0) + " wasn't in the word");
                    this.tries--;
                }
            }
            else if (guess.equals(this.word)) {
                System.out.println("You guess the word !! GG");
                break;
            } else {
                System.out.println("Not the right guess, try again");
                this.tries--;
            }

            System.out.println(this.tries + " left");
            displayWord();
        }
        hasWon();
        retry();
    }
    private void hasWon() throws IOException {
        if (this.tries > 0) {
            System.out.println("You win !");
            addScoreToFile();
        } else {
            System.out.println("You lost... better luck next time !");
        }
    }

    private void retry() throws IOException {
        Scanner in = new Scanner(System.in);
        System.out.print("Wanna retry ?: ");
        String answer = in.nextLine();
        System.out.println();
        if (answer.isEmpty() || answer.isBlank() || (!answer.equals("n")&&!answer.equals("y"))) {
            System.out.println("Enter 'y' or 'n' to answer");
            retry();
        }
        else if (answer.equals("y")) {
            reset();
            start();
        }
        else {
            System.exit(11);
        }
    }
    private void reset() {
        this.word = "";
        this.wordDisplay = "";
        this.tries = 6;
    }

    private void loadDict() throws FileNotFoundException {
        Scanner fileIn = new Scanner(new File("../minidico.txt"));
        while (fileIn.hasNextLine()) {
            this.dict.add(fileIn.nextLine());
        }
        fileIn.close();
    }

    private void selectWord() {
        int rand = this.generator.nextInt(this.dict.size());
        this.word = this.dict.get(rand);
    }
    private void initWordDisplay() {
        for (int i = 0; i < this.word.length(); i++) {
            this.wordDisplay += "_";
        }
    }

    private void printStatus() {
        System.out.println(
                "Dict: " + this.dict + '\n' +
                "DictSize: " + this.dict.size() + '\n' +
                "Word: " + this.word + '\n' +
                "WordSize: " + this.word.length() + '\n'
        );
    }
    private void displayWord() {
        for (int i = 0; i < this.wordDisplay.length(); i++) {
            System.out.print(this.wordDisplay.charAt(i));
        }
        System.out.println();
    }

    private boolean fillBlanks(Character letter) {
        String updated = "";
        for (int i = 0; i < this.wordDisplay.length(); i++) {
            if (this.word.charAt(i) == letter) {
                updated += letter;
            }
            else {
                updated += this.wordDisplay.charAt(i);
            }
        }
        this.wordDisplay = updated;
        return this.wordDisplay.contains(letter.toString());
    }

    public void addScoreToFile() throws IOException {
        String[] newFile = new String[6];
        Scanner fileIn = new Scanner(new File("../score.txt"));
        int lineCount = 0;

        while (fileIn.hasNext()) {
            String line = fileIn.nextLine();
            if (lineCount == 3) {
                String sub = line.substring(0,7);
                int score = Integer.parseInt(line.substring(7)) + 1;
                String updatedLine = sub + score;
                newFile[lineCount] = updatedLine;
            } else {
                newFile[lineCount] = line;
            }
            lineCount++;
        }
        fileIn.close();

        BufferedWriter fileOut = new BufferedWriter(new FileWriter("../score.txt"));
        for (int i = 0; i < newFile.length; i++) {
            fileOut.write(newFile[i]+'\n');
        }
        fileOut.close();
    }
}

File: Python/Hangman.py
from random import randint

class Hangman():
    def __init__(self):
        self.dict = []
        self.dict_size = 0
        
        self.word = None
        self.word_size = 0
        self.word_display = []
        
        self.tries = 6
        self.load_dict()
        
    def load_dict(self):
        file = open('../minidico.txt', 'r')
        line_count = 0
        for line in file:
            self.dict.append(line)
            line_count += 1 
        self.dict_size = line_count   
        file.close()
     
    def select_word(self):
        self.word = self.dict[randint(0, self.dict_size)]
        self.word_size = len(self.word)-1
        for i in range (0,self.word_size):
            self.word_display.append('_')
    
    def check_guess(self, guess):
        if (guess == self.word[0:len(self.word)-1]): # we remove the \n is last pos because of the file reading, tampering with results
            return True
        else:
            return False
        
    def fill_blanks(self, letter):
        modified = False
        for i in range (0, self.word_size):
            if (self.word[i] == letter):
                self.word_display[i] = letter
                modified = True
        if (not modified):
            print(letter + " wasn't in the word")
        else:
            print(letter + " was in the word, nice guess !")
        return modified
    
    def reset(self):
        self.word = None
        self.word_size = 0
        self.word_display = []
        self.tries = 6

    def print_status(self):
        print(
            "dict_size :", self.dict_size, '\n',
            "word      :", self.word,
            "word_size :", self.word_size, '\n',
            "tries left :", self.tries, '\n'
            )
        self.display_guess()
        
    def display_guess(self):
        print("".join(self.word_display))
            
    def has_won(self):
        if (self.tries <= 0):
            print("You lost, the word was, ", self.word)
            return False
        else:
            print("You won !!")
            return True
        
    def increment_score(self):
        print("+1")
        file_in = open("../score.txt", 'r')
        lines = file_in.readlines()
        
        line = lines[2]
        language = line[0:9]
        score = int(line[9:len(line)]) + 1
        lines[2] = language + str(score) +'\n'
        
        file_out = open("../score.txt", 'w')
        file_out.writelines(lines)
        
        file_in.close()
        file_out.close()
        
    def retry(self):
        x = input("Wanna play again ? (y/n): ")
        if (x == 'y'):
            self.reset()
            self.select_word()
            self.start()

    def start(self):
        self.select_word()
        
        while(self.tries > 0):
            self.print_status()
            guess = input("Make a guess : ")
            
            if (len(guess) == 1):
                if (self.fill_blanks(guess) == False):
                    self.tries -= 1
            elif (self.check_guess(guess)):
                break
            else :
                self.tries -= 1
            
        if (self.has_won()):
            self.increment_score()
        
        self.retry()
            
h = Hangman()
h.start()